src/puya/awst_build/eb/log.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import mypy.nodes
+
+from puya.awst import wtypes
+from puya.awst.nodes import (
+    BytesConstant,
+    Expression,
+    IntrinsicCall,
+    Literal,
+    ReinterpretCast,
+    UInt64Constant,
+)
+from puya.awst_build.eb.base import (
+    ExpressionBuilder,
+    IntermediateExpressionBuilder,
+)
+from puya.awst_build.eb.var_factory import var_expression
+from puya.awst_build.utils import expect_operand_wtype
+from puya.errors import CodeError
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    import mypy.types
+
+    from puya.parse import SourceLocation
+
+
+class LogBuilder(IntermediateExpressionBuilder):
+    def call(
+        self,
+        args: Sequence[ExpressionBuilder | Literal],
+        arg_kinds: list[mypy.nodes.ArgKind],
+        arg_names: list[str | None],
+        location: SourceLocation,
+        original_expr: mypy.nodes.CallExpr,
+    ) -> ExpressionBuilder:
+        args_ = list(args)
+        try:
+            sep_index = arg_names.index("sep")
+        except ValueError:
+            sep: Expression = BytesConstant(value=b"", source_location=location)
+        else:
+            sep = expect_operand_wtype(args_.pop(sep_index), wtypes.bytes_wtype)
```

Comment:
this will need to handle `str` literals, since it's allowed in your stub definition

Reply:
Done